gobblin-service/src/main/java/org/apache/gobblin/service/modules/flow/MultiHopFlowCompiler.java

```diff
@@ -87,12 +89,13 @@ public MultiHopFlowCompiler(Config config, Optional<Logger> log) {
 
   public MultiHopFlowCompiler(Config config, Optional<Logger> log, boolean instrumentationEnabled) {
     super(config, log, instrumentationEnabled);
-    this.flowGraph = new BaseFlowGraph();
-    Optional<FSFlowTemplateCatalog> flowCatalog = Optional.absent();
+    ReadWriteLock rwLock = new ReentrantReadWriteLock(true);
+    this.flowGraph = new BaseFlowGraph(rwLock);
```

Comment:
Using the RW lock of the FlowGraph works, but may not be desirable. The semantics of the RW lock of the BaseFlowGraph is to control read-write access to the FlowGraph. It seems that it might be better to have a different RW lock for controlling access to the FlowEdgeTemplateCatalog. The compiler can acquire this lock at the beginning of compileFlow() and release it once the compilation is done.